tensorflow/compiler/jit/xla_compilation_cache.cc

```diff
@@ -335,12 +346,128 @@ void LogOnceXlaCompiledFirstCluster() {
+  {
+    mutex_lock lock(cluster_compile_stats_mu_);
+    auto it = cluster_compile_stats_.find(function_name);
+    const uint64 compile_time_s = compile_time_us / 1.0e6;
+    it->second.compile_count++;
+    it->second.cumulative_compile_time_us += compile_time_us;
+
+    LogOnceXlaCompiledFirstCluster();
+    VLOG(1) << "compiled " << function_name << " "
+            << it->second.compile_count
+            << " times, compile time: " << compile_time_us
+            << " us, cumulative: " << it->second.cumulative_compile_time_us
+            << " us ("
+            << tensorflow::strings::HumanReadableElapsedTime(compile_time_s)
+            << " / "
+            << tensorflow::strings::HumanReadableElapsedTime(
+                   it->second.cumulative_compile_time_us / 1.0e6)
+            << ")";
+
+    XlaJitCompilationActivity jit_compilation_activity;
+    jit_compilation_activity.set_cluster_name(function_name);
+    jit_compilation_activity.set_compile_count(it->second.compile_count);
+    jit_compilation_activity.set_compile_time_us(compile_time_us);
+    jit_compilation_activity.set_cumulative_compile_time_us(
+        it->second.cumulative_compile_time_us);
+    TF_RETURN_IF_ERROR(
+        BroadcastXlaActivity(std::move(jit_compilation_activity)));
+  }
+
+  return Status::OK();
+}
+
+Status XlaCompilationCache::CompileAsynchronous(
+    Entry* entry, const XlaCompiler::Options& options,
+    const std::vector<XlaCompiler::Argument>& args,
+    const string &function_name,
+    const std::function<Status(XlaCompiler* compiler,
+                               const std::vector<XlaCompiler::Argument>& args,
+                               XlaCompiler::CompilationResult*)>& compile_fn) {
+  // Explicitly capture all required data by value for async compilation.
+  struct AsyncCompileData {
+    Entry* entry;
+    const XlaCompiler::Options& options;
+    const std::vector<XlaCompiler::Argument>& args;
+    const string &function_name;
+    const std::function<Status(XlaCompiler* compiler,
+                               const std::vector<XlaCompiler::Argument>& args,
+                               XlaCompiler::CompilationResult*)>& compile_fn;
+  } async_compile_data = {
+    entry, options, args, function_name, compile_fn
+  };
+
+  entry->compile_state = CompileState::kCompiling;
+  {
+    mutex_lock lock(async_compilation_.async_compilation_mu_);
+    async_compilation_.num_ongoing_compilations++;
+  }
+  // Don't move the above code into the thread function!!!
+
+  // Do not add additional variables to the capture list,
+  // but add any additional ones to the AsyncCompileData struct above
+  // When the ThreadPool for the compilation cache is destroyed, it waits for
+  // compilations to have finished. This means that 'async_compile_data.entry'
+  // and 'this' will be alive for the duration of the compilation.
+  async_compilation_.compiler_threads.get()->Schedule(
+    [this, async_compile_data] {
```

Comment:
`this` should also be in `async_compile_data`.  Same question as above about `operator()`.

Reply:
here as well, the code is incorrect.
When the struct has non-reference members, values are copied twice.\
reverting to [=] capture with a comment.